src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -1097,13 +1176,40 @@ impl<'tcx> LocationTree {
                 alloc_id,
                 span,
                 visit_children,
-            )
+            )?)
         } else {
             // `SkipChildrenOfAccessed` only gets set on protector release.
             // Since a wildcard reference are never protected this assert shouldn't fail.
             assert!(matches!(visit_children, ChildrenVisitMode::VisitChildrenOfAccessed));
+            None
+        };
+
+        let accessed_root_tag = accessed_root.map(|idx| nodes.get(idx).unwrap().tag);
+        if matches!(visit_children, ChildrenVisitMode::SkipChildrenOfAccessed) {
+            // FIXME: approximate which roots could be children of the accessed node and only skip them instead of all other trees.
+            return interp_ok(());
+        }
+        for root in roots {
+            // We don't perform a wildcard access on the tree we already performed a
+            // normal access on.
+            if Some(root) == accessed_root {
+                continue;
+            }
+            // This can only be a local access for nodes that are a parent of the accessed node.
+            // We could pass access_source as `max_local_tag`, but using accessed_root is better
+            // since that will be smaller, and it is still a node that must be
+            // on all parent paths starting at the accessed node.
+            //
+            // If we accessed the main tree then its root will be the smallest tag in the entire
+            // allocation, meaning we perform only a foreign access on all other subtrees.
```

Comment:
```suggestion
            // The choice of `max_local_tag` requires some thought.
            // This can only be a local access for nodes that are a parent of the accessed node
            // and are therefore smaller, so the accessed node itself is a valid choice for `max_local_tag`.
            // However, using `accessed_root` is better since that will be smaller. It is still a valid choice
            // because for nodes *in other trees*, if they are a parent of the accessed node then they
            // are a parent of `accessed_root`.
            //
            // If we accessed the main tree then its root will be the smallest tag in the entire
            // allocation, meaning we perform only a foreign access on all other subtrees.
```